File: 0001-0500/0038/0038.count-and-say.js
/*
 * @lc app=leetcode id=38 lang=javascript
 *
 * [38] Count and Say
 */

// @lc code=start
/**
 * @param {number} n
 * @return {string}
 */
var countAndSay = function (n) {
    let result = '1';
    for (let i = 1; i < n; i++) {
        result = say(result);
    }
    return result;
};

var say = function (str) {
    let result = '';
    let count = 1;
    let current = str[0];
    for (let i = 1; i < str.length; i++) {
        if (str[i] === current) {
            count++;
        } else {
            result += count + current;
            count = 1;
            current = str[i];
        }
    }
    result += count + current;
    return result;
}
// @lc code=end

File: 0001-0500/0038/0038.count-and-say.cs
/*
 * @lc app=leetcode id=38 lang=csharp
 *
 * [38] Count and Say
 */

// @lc code=start
public class Solution {
    public string CountAndSay(int n) {
        if (n == 1) {
            return "1";
        }
        string prev = CountAndSay(n - 1);
        StringBuilder sb = new StringBuilder();
        int count = 1;
        for (int i = 0; i < prev.Length; i++) {
            if (i + 1 < prev.Length && prev[i] == prev[i + 1]) {
                count++;
            } else {
                sb.Append(count);
                sb.Append(prev[i]);
                count = 1;
            }
        }
        return sb.ToString();   
    }
}
// @lc code=end

